File: src/engine/score-recipe.ts
import { bucketFor } from '@/engine/bucket';
import type { IngredientId, Minutes, Recipe, ScoredRecipe, UserPreferences } from '@/engine/types';

/**
 * Scoring weights.
 *
 * The spec declares `ScoredRecipe.score` but never defines its terms
 * (Technical Spec §4.1), and the AI playbook says to write this by
 * hand. These weights are reasoned, not fitted — revisit after the Aug 18-22
 * user-testing round.
 *
 * Bucketing is by missing count, so score only orders recipes WITHIN a bucket.
 *
 * Coverage dominates because it is the user's actual question. The skip penalty
 * outweighs a marginal coverage gain because an explicit skip is the only
 * negative signal the user volunteers; a disliked recipe never reaches scoring
 * at all, since `decide` eliminates it outright.
 */
export const WEIGHTS = {
  coverage: 0.6,
  timeFit: 0.25,
  cuisineMatch: 0.15,
  calorieGoal: 0.1,
  skipPenalty: 0.3,
} as const;

export function scoreRecipe(
  recipe: Recipe,
  pantry: ReadonlySet<IngredientId>,
  prefs: UserPreferences,
  timeLimit: Minutes
): ScoredRecipe {
  const missing = recipe.ingredients.filter((ing) => !pantry.has(ing.id)).map((ing) => ing.id);

  const total = recipe.ingredients.length;
  const coverage = total === 0 ? 0 : (total - missing.length) / total;

  // Clamped so a recipe surfaced by time relaxation cannot drive the score
  // arbitrarily negative.
  const timeFit = timeLimit <= 0 ? 0 : clamp01(1 - recipe.totalTimeMinutes / timeLimit);

  const cuisineMatch =
    prefs.preferredCuisine !== null && recipe.cuisine === prefs.preferredCuisine ? 1 : 0;

  const skipped = prefs.skippedRecipeIds.has(recipe.id) ? 1 : 0;
  const calorieGoal = calorieGoalFit(recipe, prefs);

  const score =
    WEIGHTS.coverage * coverage +
    WEIGHTS.timeFit * timeFit +
    WEIGHTS.cuisineMatch * cuisineMatch +
    WEIGHTS.calorieGoal * calorieGoal -
    WEIGHTS.skipPenalty * skipped;

  return { recipe, missing, bucket: bucketFor(missing.length), score };
}

/**
 * Calories are a soft preference only. Missing or low-confidence nutrition
 * never removes a recipe and contributes no ranking signal.
 */
function calorieGoalFit(recipe: Recipe, prefs: UserPreferences): number {
  const energy = recipe.energyKcalPerServing;
  if (
    (recipe.nutritionConfidence !== 'high' && recipe.nutritionConfidence !== 'medium') ||
    energy === null ||
    !Number.isFinite(energy) ||
    energy <= 0
  )
    return 0;

  // Missing and explicit null preferences are neutral and leave standard ranking.
  const goal = prefs.bodyGoal ?? null;
  if (goal === null || goal === 'maintain') return 0;

  const normalized = Math.min(1, Math.max(0, energy / 1_000));
  return goal === 'lose' ? 1 - normalized : normalized;
}

function clamp01(n: number): number {
  return Math.min(1, Math.max(0, n));
}

File: HomeChef/Engine/ScoreRecipe.swift
// Scoring weights are reasoned, not fitted — revisit after user-testing.
//
// Bucketing is by missing count, so score only orders recipes WITHIN a bucket.
// Coverage dominates because it is the user's actual question. The skip
// penalty outweighs a marginal coverage gain because an explicit skip is
// the only negative signal the user volunteers; a disliked recipe never
// reaches scoring at all, since decide() eliminates it outright.

import Foundation

enum ScoringWeights {
    static let coverage:     Double = 0.6
    static let timeFit:      Double = 0.25
    static let cuisineMatch: Double = 0.15
    static let skipPenalty:  Double = 0.3
}

func scoreRecipe(
    recipe: Recipe,
    pantry: Set<IngredientId>,
    prefs: UserPreferences,
    timeLimit: Minutes
) -> ScoredRecipe {
    let missing = recipe.ingredients
        .filter { !pantry.contains($0.id) }
        .map { $0.id }

    let total = recipe.ingredients.count
    let coverage: Double = total == 0 ? 0 : Double(total - missing.count) / Double(total)

    // Clamped so a recipe surfaced by time relaxation cannot drive the score
    // arbitrarily negative.
    let timeFit: Double = timeLimit <= 0
        ? 0
        : clamp01(1 - Double(recipe.totalTimeMinutes) / Double(timeLimit))

    let cuisineMatch: Double =
        prefs.preferredCuisine != nil && recipe.cuisine == prefs.preferredCuisine ? 1 : 0

    let skipped: Double = prefs.skippedRecipeIds.contains(recipe.id) ? 1 : 0

    let score = ScoringWeights.coverage     * coverage
              + ScoringWeights.timeFit      * timeFit
              + ScoringWeights.cuisineMatch * cuisineMatch
              - ScoringWeights.skipPenalty  * skipped

    return ScoredRecipe(
        recipe: recipe,
        missing: missing,
        bucket: bucketFor(missingCount: missing.count),
        score: score
    )
}

private func clamp01(_ n: Double) -> Double {
    min(1, max(0, n))
}
